es private fields in in

acutmore opened the pull request:
*Issue number of the reported bug or feature request: *
https://github.com/microsoft/TypeScript/issues/42574

**Describe your changes**
Implements the stage 3 'private-fields-in-in' proposal

https://github.com/tc39/proposal-private-fields-in-in

## WIP Progress

- [x] Add new 'PrivateIdentifierInInExpression' expression
    - `'prop' in obj` is a BinaryExpression, but this isn't because syntactically the left side can only be a `PrivateIdentifier`
    - suggestions for alternative names instead of 'PrivateIdentifierInInExpression' ?
- [x] Error when the `PrivateIdentifier` can't be looked-up symbolically
    - outside of a class
    - no containing class with matching private field
- [x] Expression returns `boolean`
- [x] RHS must be object or any (and not null in strictNullChecks)
    - i.e. same rules as `'prop' in obj`
- [ ] Add new error messages instead of piggybacking on existing diagnostics
- [x] Use in flowNodes and type narrowing
- [x] language server hints
    - suggest completion for privateFields on typing `#`
- [x] emit ESNext
- [x] emit down-level
- [x] part of find references results
- [x] private static field/method compatibility

acutmore committed:
basic parsing of private-id-in-in

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
basic type check and narrow

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
narrow to instanceOf klass not typeof klass

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
tweaking tests

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
fix emit with type syntax on rhs

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
add test case for subclass union narrow

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
es2020 emit

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
fix linting errors

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
mark as used

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -22750,6 +22750,26 @@ namespace ts {
+                // TODO(aclaymore): verify assertion is valid
+                Debug.assert(ctorSigs.length > 0, "narrowTypeByPrivateIdentifierInInExpression should always find the class signature");
+                const instanceType = getReturnTypeOfSignature(ctorSigs[0]);
+                return getNarrowedType(type, instanceType, assumeTrue, isTypeSubtypeOf);
```
I wonder if the `isTypeDerivedFrom` relationship is not more appropriate here (as it is used in `narrowTypeByInstanceof`). I am not 100% if  the difference between `isTypeDerivedFrom`  and `isTypeSubtypeOf` affects us here. From looking at the code `isTypeDerivedFrom` is a simpler check (just looks up the inheritance chain) while `isTypeSubtypeOf` is a full structural subtype check, so I think we should prefer `isTypeDerivedFrom`

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -26265,6 +26287,17 @@ namespace ts {
             }
         }
 
+        function lookupClassForPrivateIdentifierDeclaration(id: PrivateIdentifier): ClassLikeDeclaration | undefined {
```
I don't feel right about the duplication of `lookupSymbolForPrivateIdentifierDeclaration` this does. Maybe we can change and merge these two functions.  Or just use `lookupSymbolForPrivateIdentifierDeclaration` as I think `prop?.parent` will be the symbol for the class. If you have the symbol you won't need an extra `getSymbolOfNode` which is what happens at line `22766` anyway.

dragomirtitian commented on src/compiler/parser.ts:
```diff
@@ -4384,9 +4387,26 @@ namespace ts {
+            const id = parsePrivateIdentifier();
+            if (token() !== SyntaxKind.InKeyword) {
+                // TODO(aclaymore) use better error
+                return createMissingNode(SyntaxKind.InKeyword, /*reportAtCurrentPosition*/ true, Diagnostics._0_expected, tokenToString(SyntaxKind.InKeyword));
```
Maybe we should think of common parsing mistakes like `if(#prop.a in foo)`. This currently issues the error  `'in' expected.`. I would argue an error along the lines of: You might have meant `this.#props.a` would be better,

acutmore replied:
Looking into this there are existing errors and quickFixes for 'bare' privateFields but my branch was preventing them from running as I was consuming the field in the parser. I've changed the parser to lookahead and only match when the `in` keyword is present.

dragomirtitian commented on src/compiler/transformers/classFields.ts:
```diff
@@ -133,6 +135,36 @@ namespace ts {
+                const receiver = visitNode(node.expression, visitor, isExpression);
+
+                // TODO(aclaymore): Should this be abstracted into a factory function?
+                return setOriginalNode(
```
I think abstracting to a function might be a good idea. This is the pattern for most access to private fields. I knwo babel does not follow the same pattern for `#in`, I am not sure why, but I would consider doing so.

acutmore replied:
Cool thanks. I still haven't done this in the most recent push but have left the TODO

dragomirtitian commented on tests/baselines/reference/privateNameInInExpression.errors.txt:
```diff
@@ -0,0 +1,139 @@
+    
+            const d = #p1 in (v as Foo); // Good d is boolean (not true)
+    
+            const e = #p1 in (v as unknown); // Bad - RHS of in must be object type or any
```
I would like to see `never` added to the tests here

acutmore replied:
Test added. `'prop' in (v as never)` returns `boolean` rather than `never` so assuming that is what we'll do too

dragomirtitian commented on tests/baselines/reference/privateNameInInExpression.errors.txt:
```diff
@@ -0,0 +1,139 @@
+            }
+    
+            if (#p1 in fs) {
+                fs; // good fb is Foo (or FooSub?)
```
I would expect this to be `FooSub`. The starting type for `fs` matters, much like `if(fs instanceof Foo)` narrows `fs` to FooSub. The issue here is that `FooSub` is structurally identical to `Foo` so TS gets a bit confused.

acutmore replied:
Good spot, I was wondering why the type hadn't narrowed the way I expected. I've made `FooSub` more distinct now which gives the right type.

dragomirtitian commented on tests/baselines/reference/privateNameInInExpression.errors.txt:
```diff
@@ -0,0 +1,139 @@
+                fs; // good fs is never
+            }
+    
+            if (#p1 in b) {
+                b; // good b is 'Bar & Foo'
+            } else {
+                b; // good b is Bar
+            }
```
I wonder if we shouldn't be more strict here. I mean for `b instanceof Foo` it makes sense since you have `Symbol.species` which can customize how `instanceof` works. But for `#p1 in b` there is not universe in which `#p1` can be in `b` and  it also be an instance of `Bar` (if `Bar` is a class). For other object types this still makes sense... so I'm rather torn if we should issue an error or not.

acutmore replied:
Interesting. I'm thinking not unless the rule is applied more generally, as in all productions of `ClassA & ClassB` would be `never` if at least one class had private fields, and the other was not a sub/superType.

Surprisingly there are cases where the runtime type here could be `Bar` and not `Foo`, though I wouldn't expect this to be common code:

```js
class SuperReturn {
  constructor(v) { return v; }
}
class C extends SuperReturn {
  #priv;
  static isC(v) { return #priv in v }
  cMethod() {}
}
class Unrelated {
  random = 1;
  uMethod();
}

const u = new C(new Unrelated));
if (C.isC(u)) {
  u.random; // 1
  u.uMethod(); // fine
  u.cMethod(); // throws
}
```

dragomirtitian commented on tests/baselines/reference/privateNameInInExpressionUnused.errors.txt:
```diff
@@ -0,0 +1,18 @@
+    // TODO(aclaymore): verify we want this behavior
+    
+    class Foo {
+        private readonly unused: undefined; // expect unused error
```
Why is `unused` relevant to the test .

acutmore replied:
Just here to confirm that the test is setup correctly to report unused code, as tests by default don't. Otherwise it could silently fail. I've changed the `unused` prop to be a closer match by changing it to also be a privateField.

acutmore committed:
improve error message when inside a class

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
respect whitespace and comments in emit

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
dont parse single privateIdentifier to reenable quick fix

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
remove need for lookupClassForPrivateIdentifierDeclaration

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
add tests for never

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
give child class different structure from parent

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
cleanup testing for unused field

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
use isTypeDerivedFrom

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
add TODO for static private field emit

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore commented:
Another thing I've noticed is that `#priv in v` is not coming up in 'find all references' searches. So I need to fix that too.
EDIT: done

acutmore committed:
include in find-all-references results

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
emit as helper with runtime type check

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
correct issues caused by rebase

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
update emit for static fields and methods

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
narrow to class for static field

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
tidy up TODOs

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
add quickFix

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
update typescript.d.ts baseline

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore commented:
@dragomirtitian - I've rebased onto TypeScript post the private methods & static fields PR. And ready for another round of review when you have the time.

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -23232,6 +23232,33 @@ namespace ts {
+                }
+                const classSymbol = symbol.parent!;
+                const classType = <InterfaceType>getTypeOfSymbol(classSymbol);
+                const firstDecl = symbol.declarations?.[0];
```
Wasn't there a `valueDeclaration`  that is what is needed here ?

acutmore replied:
ah yes there is `symbol.valueDeclaration`  - that looks much better. I'll change to that.

acutmore replied:
done :)

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -23232,6 +23232,33 @@ namespace ts {
+                const firstDecl = symbol.declarations?.[0];
+                Debug.assert(firstDecl, "should always have a declaration");
+                let targetType: Type;
+                if (hasSyntacticModifier(firstDecl, ModifierFlags.Static)) {
```
use `hasStaticModifier` instead

acutmore replied:
done :)

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -23232,6 +23232,33 @@ namespace ts {
+                else {
+                    const ctorSigs = getSignaturesOfType(classType, SignatureKind.Construct);
+                    Debug.assert(ctorSigs.length > 0, "should always have a constructor");
+                    targetType = getReturnTypeOfSignature(ctorSigs[0]);
```
There seems to be a better way to get the instance type: 

```ts
const classSymbol = getSymbolOfNode(classDecl);
const classInstanceType = <InterfaceType>getDeclaredTypeOfSymbol(classSymbol);
```

At line (24161, fn: `classDeclarationExtendsNull`)

Not sure if 100% equivalent or if better, but please consider it.

acutmore replied:
done. Much better, thanks!

dragomirtitian commented on tests/baselines/reference/privateNameInInExpression.errors.txt:
```diff
@@ -0,0 +1,143 @@
+tests/cases/conformance/classes/members/privateNames/privateNameInInExpression.ts(3,27): error TS2805: Static fields with private names can't have initializers when the '--useDefineForClassFields' flag is not specified with a '--target' of 'esnext'. Consider adding the '--useDefineForClassFields' flag.
```
Nit: Consider removing this error from the test by changing compiler settings. It is noise here IMO

acutmore replied:
done :)

dragomirtitian commented on tests/baselines/reference/privateNameInInExpression.errors.txt:
```diff
@@ -0,0 +1,143 @@
+    
+            const e = #field in (v as never);
+    
+            for (let f in #field in v as any) { /**/ } // unlikely but valid
```
Took a second to get this was actually `for (let f in (#field in v as any))` 😅.

acutmore replied:
haha! I hope no one ever does this but at least we support it if they try

mkubilayk commented on src/compiler/utilities.ts:
```diff
@@ -3579,6 +3580,9 @@ namespace ts {
                         return getBinaryOperatorPrecedence(operatorKind);
                 }
 
+            case SyntaxKind.PrivateIdentifierInInExpression:
```
nit: This can be combined with `SyntaxKind.AsExpression` case.

acutmore replied:
Done :)

mkubilayk commented on src/compiler/visitorPublic.ts:
```diff
@@ -793,6 +793,12 @@ namespace ts {
 
+            case SyntaxKind.PrivateIdentifierInInExpression:
+                return factory.updatePrivateIdentifierInInExpression(<PrivateIdentifierInInExpression>node,
+                    nodeVisitor((<PrivateIdentifierInInExpression>node).name, visitor, isExpression),
```
Should this use `isPrivateIdentifier` check?

```suggestion
                    nodeVisitor((<PrivateIdentifierInInExpression>node).name, visitor, isPrivateIdentifier),

```

acutmore replied:
If my memory serves correctly: no. As I believe this check is also done on the transformed AST where `PrivateIdentifierInInExpression` becomes a methodCall.

This was over a month ago so I may be getting this mixed up with a different bit of code. I'll try it out again and confirm either way.

acutmore replied:
Played around to jog my memory. The reason this was `isExpression` and not `isPrivateIdentifier` was because when the identifier is invalid (e.g. used outside of a class) the classFields transformer emits an empty plain Identifier.

I've changed this to `isMemberName` now to match both private and standard identifiers.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -200,6 +202,28 @@ namespace ts {
+
+            // Private name has not been declared. Subsequent transformers will handle this error
+            // TODO(aclaymore): confirm this is how we want to handle the error
+            return visitEachChild(node, visitor, context);
```
This looks correct to me.

mkubilayk commented on tests/baselines/reference/privateNameInInExpressionTransform.js:
```diff
@@ -0,0 +1,97 @@
+        'prop' in v;
+        10;
+        __classPrivateFieldIn(v, _Foo_field);
+        10;
```
This looks like an existing problem but it's weird that compiler produces valid JavaScript from invalid input. It's a hazard when `noEmitOnError = false`.

```js
// in
"foo" in this = 10;

// out
"foo" in this;
10;
```

acutmore replied:
agreed. I was going for consistency with how 'prop in val' works but certainly doesn't seem ideal. I could raise an issue on the main TypeScript repo for the existing problem to gauge their opinion. What do you think?

mkubilayk replied:
A separate issue sounds good.

acutmore replied:
Issue raised https://github.com/microsoft/TypeScript/issues/44232

mkubilayk reviewed (commented):
Tried it out in VSCode, this is looking good! There is one case I found a bit inconvenient but I'm not sure if we can improve it. When you are inside `PrivateIdentifierInInExpression` and editing the identifier:

```ts
if (# in this) {
     ^ cursor
}
```

available private identifiers are listed correctly in the suggestions (`ctrl+space`). But when you pick one if them, you get a property access (`this.#foo`). If we could use the contextual information to drive this and insert `#foo` instead, that would be great. If you think people might want `this.#foo` more often than `#foo` in this case, please ignore my suggestion.

acutmore commented:
> Tried it out in VSCode, this is looking good! There is one case I found a bit inconvenient but I'm not sure if we can improve it. When you are inside `PrivateIdentifierInInExpression` and editing the identifier:
> 
> ```ts
> if (# in this) {
>      ^ cursor
> }
> ```
> 
> available private identifiers are listed correctly in the suggestions (`ctrl+space`). But when you pick one if them, you get a property access (`this.#foo`). If we could use the contextual information to drive this and insert `#foo` instead, that would be great. If you think people might want `this.#foo` more often than `#foo` in this case, please ignore my suggestion.

Yeah I was really unsure on this too. My assumption was that people are more likely to reach for accessing the private field but that's not based on any evidence just a gut feeling.

Maybe it's possible to offer both suggestions a plain `#foo` identifier when we're preceding an `in` and the `this.#foo`. I'll give it a go and see what the API supports.

acutmore committed:
post PR review improvements:

- simplify obtaining private fields's static or instance type
- make visitor node test more specific
- clean up warnings about static class fields in tests

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore commented:
I spent a little bit of time seeing if I could change the completions private fields to sometimes complete to just the identifier but couldn't get anything working quickly. I'll have another go next time I have some time.
I believe I've addressed the other comments.

acutmore committed:
Merge branch 'main' into es-private-fields-in-in

acutmore committed:
fix linting

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore committed:
update .d.ts baselines

Signed-off-by: Ashley Claymore <acutmore@users.noreply.github.com>

acutmore commented:
Merged latest TypeScript main into this branch. All tests still passing.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -201,6 +203,27 @@ namespace ts {
+         * Visits `#id in expr`
+         */
+        function visitPrivateIdentifierInInExpression(node: PrivateIdentifierInInExpression) {
+            if (!shouldTransformPrivateElements) {
```
This is probably fine today but might be a problem when `ES2022` is introduced. `shouldTransformPrivateElements` will probably be updated to say `languageVersion < ScriptTarget.ES2022`. But it's possible that [proposal-private-fields-in-in](https://github.com/tc39/proposal-private-fields-in-in) won't make it to ES2022.

acutmore replied:
The current plan is to assume https://github.com/tc39/proposal-private-fields-in-in will make stage 4 in time for ES2022. There are positive signals of this, as it is already available in latest Chrome and Firefox.

mkubilayk reviewed (approved):
I think this looks good. Left one comment to prevent against a potential future bug. 

We can use https://github.com/microsoft/TypeScript/pull/42458 as an example for the PR description when we open the upstream PR.
